lib/set.rb

```diff
@@ -151,6 +151,19 @@ def disjoint?(other)
     !intersect?(other)
   end
 
+  def divide(&block)
+    return enum_for(:divide) unless block_given?
+
+    ret = Hash.new { |h, k| h[k] = Set.new }
+    if block.arity == 2
+      values = to_a.product(to_a)
+      values.each(&block)
+    else
+      each { |val| ret[block.call(val)] << val }
+    end
+    ret.values
+  end
+
```

Comment:
Wow, this is way simpler than I ever thought it would be. I had an implementation that worked for arity 1 but not 2, but it was wayyy bigger than this. Great work.

Reply:
Just to be clear: the arity 2 version does not work yet.

Reply:
Oh ok! 😆

Reply:
It turned out to be even easier than the first implementation: There is a `Set#classify` which does pretty much all the preparation we need.

The two argument version now works as well. The code is pretty ugly though.